Sources/NIOCore/VsockAddress.swift

```diff
@@ -0,0 +1,221 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if canImport(Darwin) || os(Linux)
+
+#if canImport(Darwin)
+import Darwin
+import CNIODarwin
+fileprivate let get_local_vsock_cid: @convention(c) (CInt) -> UInt32 = CNIODarwin_get_local_vsock_cid
+#elseif os(Linux)
+#if canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#endif
+import CNIOLinux
+fileprivate let get_local_vsock_cid: @convention(c) (CInt) -> UInt32 = CNIOLinux_get_local_vsock_cid
+#else
+#error("Unable to identify your C library.")
+#endif
+
+/// A vsock socket address.
+///
+/// A socket address is defined as a combination of a Context Identifier (CID) and a port number.
+/// The CID identifies the source or destination, which is either a virtual machine or the host.
+/// The port number differentiates between multiple services running on a single machine.
+///
+/// For well-known CID values and port numbers, see ``ContextID`` and ``Port`` respectively.
+public struct VsockAddress: Sendable {
+    private let _storage: Box<sockaddr_vm>
+
+    /// The libc socket address for a vsock socket.
+    public var address: sockaddr_vm { return _storage.value }
```

Comment:
For now I think my preference is not to make this API, so we avoid committing to a non-portable representation here. That would make it easy to have this type on Windows if we need it.

Reply:
Relatedly, can I propose we create this lazily and avoid the use of `Box`? We only need `Box` for `SocketAddress` to deal with existential values, but `ContextID` and `Port` together as stored properties should fit nicely.

Reply:
Yeah, `SocketAddress` making the various `sockaddr` stuff API was a historical mistake we cannot unwind.

Yes, I was assuming (1).